relay/DCUtR.md

```diff
@@ -0,0 +1,104 @@
+# Direct Connection Upgrade through Relay
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision    |
+|-----------------|---------------|--------|--------------------|
+| 1A              | Working Draft | Active | DRAFT, 2019-05-29  |
+
+Authors: vyzo
+
+Interest Group: raulk, stebalien, whyrusleeping
+
+## Introduction
+
+NAT traversal is a quintessential problem in peer-to-peer networks.
+
+We currently utilize relays, which allow us to traverse NATs by using
+a third party as proxy. Relays are a reliable fallback, that can
+connect peers behind NAT albeit with a high-latency, low-bandwidth
+connection.  Unfortunately, they are expensive to scale and maintain
+if they have to carry all the NATed node traffic in the network.
+
+It is often possible for two peers behind NAT to communicate directly
+by utilizing a technique called _hole punching_[1]. The technique
+relies on the two peers synchronizing and simultaneously opening
+connections to each other to their predicted external address. It
+works well for UDP, with an estimated 80% success rate, and reasonably
+well for TCP, with an estimated 60% success rate.
+
+The problem in hole punching, apart from not working all the time, is
+the need for rendezvous and synchronization. This is usually
+accomplished using dedicated signaling servers [2].  However, this
+introduces yet another piece of infrastructure, while still requiring
+the use of relays as a fallback for the cases where a direct
+connection is not possible.
+
+In this draft, we describe a synchronization protocol for direct
+connectivity with hole punching that eschews signaling servers and
+utilizes existing relay connections instead.  That is, peers start
+with a relay connection and synchronize directly, without the use of a
+signaling server.  If the hole punching attempt is successful, the
+peers _upgrade_ their connection to a direct connection and they can
+close the relay connection.  If the hole punching attempt fails, they
+can keep using the relay connection as they were.
+
+## The Protocol
+
+Consider two peers, `A` and `B`. `A` wants to connect to `B`, which is
+behind a NAT and advertises relay addresses. `A` may itself be behind
+a NAT or be a public node.
+
+The protocol starts with the completion of a relay connection from `A`
+to `B`.  Upon observing the new connection, the inbound peer (here `B`)
```

Comment:
There's a chance that `A` didn't know the current, full address set of `B`, e.g. `A` had stale addresses for `B` in its local address book that didn't include public addresses that `B` learnt of itself until later. So I think this protocol needs to be bi-directional for additional resiliency.

Also because `B` could be malicious and be actively pursuing something analogous to a "downgrade attack" at the transport layer, by picking a worse-performing transport deliberately, even if it knows of a better path to `A`.

Reply:
This should be covered by the upgrade protocol itself, that's the bidirectional part.

Reply:
I had misunderstood the spec on first read. This seems OK.